events/controller.go

```diff
@@ -0,0 +1,65 @@
+package events
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/awslabs/operatorpkg/object"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/utils/clock"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+type Controller[T client.Object] struct {
+	kind       string
+	startTime  time.Time
+	kubeClient client.Client
+}
+
+func NewController[T client.Object](client client.Client, clock clock.Clock) *Controller[T] {
+	return &Controller[T]{
+		kind:       object.GVK(object.New[T]()).Kind,
+		startTime:  clock.Now(),
+		kubeClient: client,
+	}
+}
+
+func (c *Controller[T]) Register(_ context.Context, m manager.Manager) error {
+	return controllerruntime.NewControllerManagedBy(m).
+		For(&v1.Event{}, builder.WithPredicates(predicate.NewTypedPredicateFuncs(func(o client.Object) bool {
+			// Only reconcile on the object kind we care about
+			event := o.(*v1.Event)
+			return event.InvolvedObject.Kind == c.kind
+		}))).
+		WithOptions(controller.Options{MaxConcurrentReconciles: 10}).
+		Named(fmt.Sprintf("operatorpkg.%s.events", strings.ToLower(c.kind))).
+		Complete(reconcile.AsReconciler(m.GetClient(), c))
+}
+
+func (c *Controller[T]) Reconcile(ctx context.Context, event *v1.Event) (reconcile.Result, error) {
+	// We check if the event was created in the lifetime of this controller
+	// since we don't duplicate metrics on controller restart or lease handover
+	if c.startTime.Before(event.LastTimestamp.Time) {
+		groupVersion, err := schema.ParseGroupVersion(event.InvolvedObject.APIVersion)
```

Comment:
`lo.Must`? Do we ever expect this to fail? So what's the value of requeueing?

Reply:
I think we can treat this as an invariant and expect the developer to have passed in an object with a valid APIVersion